clippy_lints/src/methods/format_collect.rs

```diff
@@ -0,0 +1,33 @@
+use super::FORMAT_COLLECT;
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::macros::is_format_macro;
+use clippy_utils::macros::root_macro_call_first_node;
+use clippy_utils::ty::is_type_lang_item;
+use rustc_hir::Expr;
+use rustc_hir::ExprKind;
+use rustc_hir::LangItem;
+use rustc_lint::LateContext;
+use rustc_span::Span;
+
+fn tail_expr<'tcx>(expr: &'tcx Expr<'tcx>) -> Option<&'tcx Expr<'tcx>> {
```

Comment:
yep, the important bit is the check that it doesn't come from an expansion. If we always remove blocks, we would also remove parts of the `format!` expansion, which makes code further below not work (`root_macro_call_first_node` will return None, just tested it)

Reply:
I think we could check the syntax context (`.ctxt()`) of the span, so we only ignore blocks if it's from `format!` but that's a *major* nit

Also, this should probably be recursive so something like `{{ format!("{x:?}) }}` works (which I believe doesn't?)

Reply:
The name is also a bit odd as well, no suggestions though. Perhaps documenting why it isn't using `peel_blocks` could be nice

Reply:
> I think we could check the syntax context (`.ctxt()`) of the span, then `.outer_expn_data()`, so we only ignore blocks if it's from `format!` but that's a _major_ nit (There also may be a better way to do that, no idea)

I'm not sure. I think doing it like that will have FPs (unless I misunderstood) because `outer_expn_data()` returns the "outermost"(?) expansion, not the very first. In other words, it would trigger here:
```rs
macro_rules! x {
    ($b:expr) => {
        format!("{:02X}", $b)
    };
}
fn hex_encode(bytes: &[u8]) -> String {
    bytes.iter().map(|b| x!(b)).collect()
}
```
when it shouldn't, because `format!` comes from another macro expansion

> Also, this should probably be recursive so something like `{{ format!("{x:?}) }}` works (which I believe doesn't?)

Good point.

> The name is also a bit odd as well, no suggestions though. Perhaps documenting why it isn't using `peel_blocks` could be nice

Are you referring to the name of the function or the lint? I agree with both though 😅

Reply:
> I'm not sure. I think doing it like that will have FPs (unless I misunderstood) because outer_expn_data() returns the "outermost"(?) expansion, not the very first. In other words, it would trigger here:

Is that so? That's unfortunate then

> Are you referring to the name of the function or the lint? I agree with both though sweat_smile

Function, the lint's name is fine imo

Reply:
Oh I think this could use `marks` or `remove_mark` on `Span::ctxt` instead,  but I highly doubt this is worth the effort ^^ This won't change the actual outcome afaik